### ISSUE TITLE
Request: Ability to disable rules inline

### ISSUE THREAD

**Issue by dcwither** [sider/goodcheck#74]
One thing that would be really useful is to be able to disable rules inline.

This is often a necessary evil for adopting a new linter or lint rule that might catch issues in existing code, in order to prevent the addition of more problems.

This might not fit into what Goodcheck does since it searches for regex patterns while ESLint and RuboCop use AST traversal, but I figured I'd raise it.

I've thought of two possible solutions to this issue.

## Solution 1: `goodcheck-disable-line`
ESLint and RuboCop do this through disable file, line, and next line comments, e.g.

```js
/* eslint-disable */

alert('foo');
```

```js
alert('foo'); // eslint-disable-line
```

```js
/* eslint-disable-next-line */
alert('foo');
```

Sources: 
- [ESLint disabiling rules](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments) 
- [RuboCop disabling rules](https://github.com/rubocop-hq/rubocop/blob/master/manual/configuration.md#disabling-cops-within-source-code)

### Problems

- Goodcheck is seemingly language agnostic, so it won't be able to identify what's a comment.
- Hard to implement without AST analysis

### Mitigations
Only do the `disable-next-line` one, and ignoring what the comment identifier is (or having a list of them per file type)

## Solution 1: `goodcheck.ignore`

An alternative extension might be to have a file that tracks existing errors that we've accepted and compare that when running.

This would be similar to [`brakeman.ignore`](https://brakemanscanner.org/docs/ignoring_false_positives/)

## Contributions

If either of these solutions seems worthwhile, I'd also be happy to try making a PR. Any guidance is welcome since I haven't worked much with ruby codebases.

Also thanks for the incredibly fast response on https://github.com/sider/goodcheck/issues/72!

**Comment by ybiquitous** [comment]
Thank you for the request! I also think this feature is very useful. 😊 

I prefer the solution *1* and the following mitigation makes sense to me.

> Only do the disable-next-line one, and ignoring what the comment identifier is (or having a list of them per file type)

Goodcheck is based on Regexp, so I don't want to handle AST (keep it simple).
Then, we need some logic detecting a file type (from a file extenstion, shebang, or use a library?)

For example:

```ruby
# Ruby example

# goodcheck-disable-next-line
foo "bar"
```

```js
// JavaScript example

// goodcheck-disable-next-line
foo("bar);
```

```markdown
<!--- Markdown example --->

<!--- goodcheck-disable-next-line --->
foo *bar*
```

---

> If either of these solutions seems worthwhile, I'd also be happy to try making a PR. Any guidance is welcome since I haven't worked much with ruby codebases.

Thank you for the offer. I think you will modify the following `Goodcheck::Analyzer#scan` method:

https://github.com/sider/goodcheck/blob/389c7f266b19087ccc2dfceaaf706b28c6fb526c/lib/goodcheck/analyzer.rb#L13-L36

**Comment by dcwither** [comment]
Glad you think this could be worthwhile. Thanks for pointing me in the right direction! I'm sick right now, but when I'm feeling better I'll pull down the repo, and do some digging.

I'll also try to write some appropriate tests.

**Comment by ybiquitous** [comment]
FYI. This is a Prettier way:
https://prettier.io/docs/en/ignore.html

**Comment by dcwither** [comment]
Thanks, I've done some looking and I found that the scanner cannot easily step backwards, and checking the pre-match block would be awkward since there's no easy/fast way I can think of to find the "previous line" using the scanner.

My current assessment is that this should be another module in addition to the analyzer. 

What this would do is:

1. find all the ranges that are `disabled` 
2. filter the issues by which ones don't overlap with the disabled ranges

Does this seem like a reasonable path?

The unfortunate update is I've been told that because this project is licensed under AGPL 3.0 (https://github.com/sider/goodcheck/issues/45) we can't use if for our purposes, so I won't be able to devote more work hours to this. I'm still interested in working on this but it'll be slower due to being limited by my personal time.

**Comment by ybiquitous** [comment]
@dcwither We have updated the license to MIT. We will release a new version including the license change soon.

About our answer to your assessment of the implementation, please wait for a while. 🙏

**Comment by ybiquitous** [comment]
The version 2.4.3 has been released! 🎉 
https://github.com/sider/goodcheck/releases/tag/v2.4.3

**Comment by dcwither** [comment]
You are all the best! I'm really excited to add to this project!

**Comment by ybiquitous** [comment]
@dcwither Sorry for my too late reply. 🙇 

> My current assessment is that this should be another module in addition to the analyzer.
> 
> What this would do is:
> 
>     1. find all the ranges that are `disabled`
> 
>     2. filter the issues by which ones don't overlap with the disabled ranges
> 
> 
> Does this seem like a reasonable path?

Your suggestion makes sense to me. I'm concerned a bit about any effect on the performance, but the idea seems very simple and good. 👍

**Comment by ybiquitous** [comment]
I've released the new version 2.5.0 including this issue.

https://github.com/sider/goodcheck/releases/tag/v2.5.0

@dcwither I really appreciate your work. Thank you!
